pkg/controller/usersignup/usersignup_controller.go

```diff
@@ -97,5 +116,189 @@ func (r *ReconcileUserSignup) Reconcile(request reconcile.Request) (reconcile.Re
 		return reconcile.Result{}, err
 	}
 
+	// Read the status conditions and assign some variables
+	provisioning := condition.IsTrue(instance.Status.Conditions, toolchainv1alpha1.UserSignupProvisioning)
+	completed := condition.IsTrue(instance.Status.Conditions, toolchainv1alpha1.UserSignupComplete)
+
+	// If the status is not yet provisioning or completed, check for automatic or manual approval
+	if !provisioning && !completed {
```

Comment:
in which cases is the reconcile triggered and this condition is not fulfilled? I would assume that every time it should try to ensure the desired state...
btw, we shouldn't rely on the status in the controllers - the status is for the clients

Reply:
A newly created `UserSignup` instance would meet these conditions.  The reason that the status values are tested here is because the provisioning process likely occurs in a separate cluster, and would only be updated once the `UserAccountController` finishes its work.

Reply:
ah, I see. The controller watches also the MUR. So when it is created then it is reconciled again. but wait, where do you set the status to completed? in the previous version of your PR it was at the end of the inner if part, but you have removed it as part of this commit https://github.com/codeready-toolchain/host-operator/pull/27/commits/fb27655f504a806e6a4625436c6fe1fc348f39dd
Btw, wouldn't be good to watch only the creation/deletion of the MUR and ignore all updates (mainly the status ones)? you can check this [predicate](https://github.com/codeready-toolchain/member-operator/blob/master/pkg/predicate/predicate.go) as an inspiration

Reply:
Good point @MatousJobanek, even I can't see where we are setting status to completed

Reply:
@MatousJobanek there was duplicate code setting the status, so I removed it.  The previous code you are referring to set the provisioning status, not complete (see [this commit](https://github.com/codeready-toolchain/host-operator/pull/27/commits/fb27655f504a806e6a4625436c6fe1fc348f39dd#diff-5b4b18a0fc2e0d708e4e6334cdb6bc69L209)) It's now set [here](https://github.com/codeready-toolchain/host-operator/pull/27/files#diff-5b4b18a0fc2e0d708e4e6334cdb6bc69R165) in the call to `r.setStatusProvisioning()`.  The `complete` status is not currently set, as it depends on the provisioning process first running successfully in the member cluster - my plan was to update this as a separate task.

Reply:
I see, so this PR should only create the MUR based on the UserSignup and the approval policy (as is also written in the description :-)). I misunderstood based on the previous changes and states of the PR, sorry for that.

From what you have written I assume that you want to wait until the whole provisioning is finished and then set the `complete` status to true, which doesn't look obvious from this diagram: https://docs.google.com/drawings/d/1f0xD4Vg0I3iZdolAHP2UlhOVdU3l3OP1z1EqZuPZ4Ew/edit?ts=5d4a0955
I thought that you wanted to just create the MUR and then set the `complete` status and let UI track the MUR.status for any failures and progress of the user account provisioning. This is also my understanding - UserSignup request should ensure that user can signup to the SaaS system (which should ensure creation of the MUR if I'm not mistaken), the rest of the user provisioning can be tracked via the MUR as it is more complex and depends on other objects in different clusters that UserSignup doesn't have a reference to. 
Just to be clear - this is only my point of view, I can be easily wrong (which happens quite often :-)), so if you have good reasoning for tracking the result of the whole provisioning through the UserSignup, then I'm fine with it ;-)

Reply:
UserSignup seems like the obvious resource from which the user may track the executive state of the user's signup process.  If we don't wish to perform any tracking at all, then there is no point having a `Provisioning` condition in the status.  Also, the `Complete` condition wouldn't really make sense either, as its meaning then becomes ambiguous - what exactly is "Complete" if the resource status doesn't reflect the state of the provisioning workflow?

You're correct about the diagram, it shouldn't have shown a termination node that sets the status to "Complete".  I've removed that for now until we make a firm decision on the design.

We could modify the workflow to have the MUR reflect the provisioning state of the user instead of UserSignup, however that will require changes to the UserSignup status conditions as I mentioned.  There is also the consideration from the UI perspective, when a new user signs up.  At signup time, then UserSignup resource is created by the UI (which then knows about the UserSignup resource because it created it) but the UI is unaware of the MasterUserRecord resource at this point in time. The MUR then may or may not be created depending on user approval policy, creation conflicts, etc, however the UI is still only aware of the UserSignup resource and if the MUR creation fails or is delayed because of the approval requirement then what should the UI do?  It only has access to UserSignup at that point in time.

Reply:
Agree with you, let's use the UserSignup for tracking the final result ;-)

Reply:
I'm still not sure we should reflect all the provisioning flow back to UserSignup resource status. We could do as an option but wouldn't have a simple flow for signup be better?

So, I see two options here:
- `UserSignup` is a wrapper for MUR (and UserAccount/NSTemplateSet/etc...) and watches/reflects the whole flow/status of user provisioning with all the complications. But in this case we will have an easier usage from UI/clients (it can watch UserSignup only to show the provisioning status).
- Use `UserSignup` purely to manage signups. So, it's not about provisioning. It's about signups/approvals. Clients/UI creates a UserSignup to get user approved. It's flow treated as complete when signup is approved (or autoapproved) and the corresponding MUR is created (just created, we don't care about its status). In this case we would have much simpler `UserSignup` flow but the clients/UI would need to watch MUR for provisioning status.

I still like the second option better. Don't really want to pull all the user provisioning complexity and status sync up to UserSignup. @sbryzak @MatousJobanek thoughts? We could discuss it the next week.

Reply:
@alexeykazakov we had a discussion about this during an impromptu bluejeans call on friday.  The issue with using MUR to track the provisioning workflow is that the MUR might not actually be the one that the UserSignupController created.  I.e. imagine you have a conflict when UserSignupController attempts to create the MUR and it already exists.  If the UI is designed to track the provisioning state via the MUR, then it will be looking at the MUR resource instance that already existed previously.  Because the UI can *only* depend on the UserSignup that it itself created, it makes most sense to me to use it to track the provisioning status.

Reply:
I'm not sure we should/can support the case when we create a MUR without Signup but it doesn't matter actually. If MUR is already exists then we just update the status to Complete:true and exit.
Here is the flow I'm suggesting - https://docs.google.com/drawings/d/1nYTN8DqRfXPP63UJr7JIBrMVqHniM_fEKmkV8E98mlA/edit?usp=sharing
Let's discuss it.